### PR TITLE
[BEAM-10672] Add Python Combine Load Tests for streaming on Flink

### PR DESCRIPTION
**Pull request by kamilwu**
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace
--- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
Build python source distribution and wheels
Python tests
Java tests

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Load Tests Python Combine Flink Batch

**Comment by kamilwu** [comment]
Run Load Tests Python Combine Flink Streaming

**Comment by kamilwu** [comment]
Run Load Tests Python Combine Dataflow Batch

**Comment by kamilwu** [comment]
Run Load Tests Python Combine Dataflow Streaming

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Load Tests Python Combine Flink Streaming

**Comment by kamilwu** [comment]
Run Load Tests Python Combine Flink Batch

**Comment by kamilwu** [comment]
Run Load Tests Python Combine Dataflow Batch

**Comment by kamilwu** [comment]
Run Load Tests Python Combine Dataflow Streaming

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Load Tests Python Combine Flink Batch

**Comment by kamilwu** [comment]
Run Load Tests Python Combine Flink Streaming

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Portable_Python PreCommit

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Load Tests Python Combine Flink Streaming

**Comment by kamilwu** [comment]
Run Load Tests Python Combine Flink Batch

**Comment by kamilwu** [comment]
Run Load Tests Python Combine Dataflow Batch

**Comment by kamilwu** [comment]
Run Load Tests Python Combine Flink Streaming

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Portable_Python PreCommit

**Comment by kamilwu** [comment]
Run Load Tests Python Combine Flink Streaming

**Comment by kamilwu** [comment]
Run Load Tests Python Combine Flink Batch

File: .test-infra/dataproc/init-actions/flink.sh [hunk 1]
```diff
@@ -160,10 +156,10 @@ jobmanager.rpc.address: ${master_hostname}
 jobmanager.heap.mb: ${flink_jobmanager_memory}
 taskmanager.memory.process.size: "${flink_taskmanager_memory} mb"
 taskmanager.memory.jvm-metaspace.size: 512 mb
+taskmanager.memory.task.off-heap.size: 256 mb
```
Comment (kamilwu) [inline .test-infra/dataproc/init-actions/flink.sh:159]: The jobs had issues with direct memory being too low. This setting solved the problem.

**Comment by kamilwu** [comment]
@mxm Would you mind taking a look?

File: sdks/python/apache_beam/testing/load_tests/load_test.py [hunk 2]
```diff
@@ -183,3 +189,66 @@ def get_option_or_default(self, opt_name, default=0):
       return type(default)(option)
     except:
       raise
+
+
+class StatefulLoadGenerator(beam.PTransform):
```
Comment (mxm) [inline sdks/python/apache_beam/testing/load_tests/load_test.py:194]: Is this the right place to put this? Should this be in `sdks/python/apache_beam/testing/synthetic_pipeline.py`?

**Review by mxm — COMMENTED** [review]
LGTM

File: sdks/python/apache_beam/testing/load_tests/load_test.py (hunk 2 above)
Comment (kamilwu) [reply sdks/python/apache_beam/testing/load_tests/load_test.py:194]: Hm, good point. I wasn't sure about that. I'll put this in `sdks/python/apache_beam/testing/synthetic_pipeline.py`

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Load Tests Python Combine Flink Streaming

**Comment by kamilwu** [comment]
Run Load Tests Python Combine Flink Streaming

**Comment by kamilwu** [comment]
Run Load Tests Python Combine Flink Streaming

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Load Tests Python Combine Flink Streaming

**Comment by kamilwu** [comment]
Publish Docker Snapshots

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Publish Docker Snapshots

**Comment by kamilwu** [comment]
Publish Docker Snapshots

**Comment by kamilwu** [comment]
Run PythonDocker PreCommit

**Comment by kamilwu** [comment]
Publish Docker Snapshots

**Comment by kamilwu** [comment]
Publish Docker Snapshots

**Comment by kamilwu** [comment]
Run Load Tests Python Combine Flink Streaming

**Comment by kamilwu** [comment]
@mxm 

I changed my approach to building Docker images. Instead of building new images every time test is executed, I think it'd be better to have a separate daily Jenkins job for that. That would bring some benefits:
* Tests would be faster
* Only 3 new images would be pushed to the registry per day, instead of 10 (5 (current number of tests) * 2 (SDK harness + job server))

Let me know what you think.

File: .test-infra/dataproc/init-actions/flink.sh (hunk 1 above)
Comment (mxm) [reply .test-infra/dataproc/init-actions/flink.sh:159]: Sounds good!

**Review by mxm — COMMENTED** [review]
Thanks @kamilwu. I think it makes sense to have a separate Jenkins job for the Docker images. Building them from master for every load test, adds a lot of overhead. Also, for reproducibility it makes sense to use the same image for all tests.

**Comment by kamilwu** [comment]
Thanks @mxm. Is there anything else to be done before merging this PR?

**Comment by mxm** [comment]
All good!

**Pull request merged**